File: backend/auth-service/api/src/commonMain/kotlin/AuthProviders.kt
package dk.sdu.cloud.auth.api

import dk.sdu.cloud.CommonErrorMessage
import dk.sdu.cloud.FindByStringId
import dk.sdu.cloud.Roles
import dk.sdu.cloud.calls.*
import kotlinx.serialization.Serializable

@Serializable
data class AuthProvidersRegisterRequestItem(val id: String)

@Serializable
data class AuthProvidersRegisterResponseItem(val claimToken: String)

@Serializable
data class PublicKeyAndRefreshToken(
    val providerId: String,
    val publicKey: String,
    val refreshToken: String,
)

typealias AuthProvidersRenewRequestItem = FindByStringId

typealias AuthProvidersRefreshRequest = BulkRequest<AuthProvidersRefreshRequestItem>
typealias AuthProvidersRefreshRequestItem = RefreshToken
typealias AuthProvidersRefreshResponse = BulkResponse<AccessToken>
typealias AuthProvidersRefreshAudit = FindByStringId

typealias AuthProvidersRetrievePublicKeyRequest = FindByStringId

@Serializable
data class AuthProvidersRetrievePublicKeyResponse(val publicKey: String)

typealias AuthProvidersRefreshAsProviderRequest = BulkRequest<AuthProvidersRefreshAsProviderRequestItem>

@Serializable
data class AuthProvidersRefreshAsProviderRequestItem(val providerId: String)
typealias AuthProvidersRefreshAsProviderResponse = BulkResponse<AccessToken>

typealias AuthProvidersGenerateKeyPairRequest = Unit
@Serializable
data class AuthProvidersGenerateKeyPairResponse(
    val publicKey: String,
    val privateKey: String,
)

@UCloudApiExperimental(ExperimentalLevel.ALPHA)
object AuthProviders : CallDescriptionContainer("auth.providers") {
    const val baseContext = "/auth/providers"
    const val PROVIDER_PREFIX = "#P_"

    @UCloudApiInternal(InternalLevel.BETA)
    val register = call<BulkRequest<AuthProvidersRegisterRequestItem>, BulkResponse<AuthProvidersRegisterResponseItem>,
        CommonErrorMessage>("register") {
        httpCreate(baseContext, roles = Roles.PRIVILEGED)
    }

    @UCloudApiInternal(InternalLevel.BETA)
    val claim = call<BulkRequest<AuthProvidersRegisterResponseItem>, BulkResponse<PublicKeyAndRefreshToken>,
        CommonErrorMessage>("claim") {
        httpUpdate(baseContext, "claim", roles = Roles.PRIVILEGED)
    }

    val renew = call<BulkRequest<AuthProvidersRenewRequestItem>, BulkResponse<PublicKeyAndRefreshToken>,
        CommonErrorMessage>("renew") {
        httpUpdate(baseContext, "renew", roles = Roles.PRIVILEGED)
    }

    val refresh = call<AuthProvidersRefreshRequest, AuthProvidersRefreshResponse, CommonErrorMessage>("refresh") {
        audit<BulkResponse<AuthProvidersRefreshAudit>>()
        httpUpdate(baseContext, "refresh", roles = Roles.PUBLIC)
    }

    @UCloudApiInternal(InternalLevel.BETA)
    val retrievePublicKey = call<AuthProvidersRetrievePublicKeyRequest, AuthProvidersRetrievePublicKeyResponse,
        CommonErrorMessage>("retrievePublicKey") {
        httpRetrieve(baseContext, "key", roles = Roles.PRIVILEGED)
    }

    @UCloudApiInternal(InternalLevel.BETA)
    val refreshAsOrchestrator = call<AuthProvidersRefreshAsProviderRequest, AuthProvidersRefreshAsProviderResponse,
        CommonErrorMessage>("refreshAsOrchestrator") {
        httpUpdate(baseContext, "refreshAsOrchestrator", roles = Roles.PRIVILEGED)

        documentation {
            summary = "Signs an access-token to be used by a UCloud service"
            description = """
                This RPC signs an access-token which will be used by authorized UCloud services to act as an
                orchestrator of resources.
            """.trimIndent()
        }
    }

    @UCloudApiInternal(InternalLevel.BETA)
    val generateKeyPair = call<AuthProvidersGenerateKeyPairRequest, AuthProvidersGenerateKeyPairResponse,
        CommonErrorMessage>("generateKeyPair") {
        httpUpdate(baseContext, "generateKeyPair", roles = Roles.PRIVILEGED)

        documentation {
            summary = "Generates an RSA key pair useful for JWT signatures"
            description = """
                Generates an RSA key pair and returns it to the client. The key pair is not stored or registered in any
                way by the authentication service.
            """.trimIndent()
        }
    }
}

File: provider-integration/shared/pkg/foundation/auth_providers.go
package foundation

import (
	"encoding/json"
	"ucloud.dk/shared/pkg/rpc"
	"ucloud.dk/shared/pkg/util"
)

const AuthProvidersContext = "auth/providers"

type AuthProviderClaimToken struct {
	ClaimToken string `json:"claimToken"`
}

type PublicKeyAndRefreshToken struct {
	ProviderId   string `json:"providerId"`
	PublicKey    string `json:"publicKey"`
	RefreshToken string `json:"refreshToken"`
}

var AuthProvidersRenew = rpc.Call[BulkRequest[FindByStringId], BulkResponse[PublicKeyAndRefreshToken]]{
	BaseContext: AuthProvidersContext,
	Operation:   "renew",
	Convention:  rpc.ConventionUpdate,
	Roles:       rpc.RolesPrivileged,
}

type RefreshToken struct {
	RefreshToken string `json:"refreshToken"`
}

var AuthProvidersRefresh = rpc.Call[BulkRequest[RefreshToken], BulkResponse[AccessTokenAndCsrf]]{
	BaseContext: AuthProvidersContext,
	Operation:   "refresh",
	Convention:  rpc.ConventionUpdate,
	Roles:       rpc.RolesPublic,
	Audit: rpc.AuditRules{
		Transformer: func(request any) json.RawMessage {
			return json.RawMessage("{}")
		},
	},
}

type FindByProviderId struct {
	ProviderId string `json:"providerId"`
}

var AuthProvidersRefreshAsOrchestrator = rpc.Call[BulkRequest[FindByProviderId], BulkResponse[AccessTokenAndCsrf]]{
	BaseContext: AuthProvidersContext,
	Operation:   "refreshAsOrchestrator",
	Convention:  rpc.ConventionUpdate,
	Roles:       rpc.RolesPrivileged,
}

// NOTE(Dan): The generate key-pair function will not be needed after provider services are migrated

type PublicAndPrivateKey struct {
	PublicKey  string
	PrivateKey string
}

var AuthProvidersGenerateKeyPair = rpc.Call[util.Empty, PublicAndPrivateKey]{
	BaseContext: AuthProvidersContext,
	Operation:   "generateKeyPair",
	Convention:  rpc.ConventionUpdate,
	Roles:       rpc.RolesPrivileged,
}

const ProviderSubjectPrefix = "#P_"
